Adding GCP Managed VM "Cloud" support.

patflynn opened the pull request:
This is the minimum set of functionality and needs to be augmented with:

1) validation
2) arbitrary war jar artifacts (i.e. the user can pick a jar or war in the file system rather than an artifact, or the user can create an artifact from an existing jar or war)
3) the destination tmp dir for staging needs to be in the git project for correct src context generation
4) currently points at gcloud path rather than cloud sdk root.
5) probably need a more obvious UI hint that files have been created when the user 'generates' their config.
6) tests?
7) Ideally we can update our config generation to use gcloud app gen-config
 managed_vm_cloud_config
 8) Ability to stop the remote service (using gcloud preview app stop
 9) A clear message on the GCP MVM Cloud Type that this costs money and does not scale to 0.
10) finally needs to use plugin's auth credentials instead of gcloud. This depends on a change on the gcloud cli side. 

 I'd like to check this in now that it's feature flag protected so Etan can iterate on it while I'm on paternity.

patflynn commented:
related to #270.

elharo commented:
Can you file the additional work as independent feature requests?

elharo commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -16,6 +16,23 @@
+cloud.project.label=Project\:
+
+appengine.cloudsdk.location.label=Cloud SDK path\:
+appengine.cloudsdk.location.browse.button=Browse to Cloud SDK location
```
Usually in these circumstances the button just says "Browse". The above label says what the user is browsing for.

elharo commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -16,6 +16,23 @@
+
+appengine.cloudsdk.location.label=Cloud SDK path\:
+appengine.cloudsdk.location.browse.button=Browse to Cloud SDK location
+appengine.appyaml.location.browse.button=Browse for Your App.Yaml Location
```
ditto

elharo commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -16,6 +16,23 @@
+appengine.managedvm.config.choose.destination.folder.button=Choose destination folder
+appengine.managedvm.config.destination.chooser.title=Choose Generated Configuration Destination Folder
+appengine.managedvm.generate.targetdirectory.label=Choose a destination directory
+appengine.generate.config.button=Generate..
```
three dots

elharo commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -16,6 +16,23 @@
+appengine.managedvm.configtype.custom.label=Custom
+appengine.managedvm.dockerfile.location.label=Dockerfile path\:
+appengine.managedvm.config.destination.folder.label=Choose destination folder\:
+appengine.managedvm.config.choose.destination.folder.button=Choose destination folder
```
ditto

elharo commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineHelper.java:
```diff
@@ -0,0 +1,84 @@
+  /**
+   * Creates a {@link Runnable} that will perform custom MVM deployment on {@code run()).
+   *
+   * @param loggingHandler logging messages will be outputted to this
```
outputted --> output

elharo commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineHelper.java:
```diff
@@ -0,0 +1,84 @@
+   * Creates a {@link Runnable} that will perform a standard MVM deployment with an automatically
+   * configured runtime (app.yaml and Dockerfile) on {@code run()).
+   *
+   * @param loggingHandler logging messages will be outputted to this
```
output

elharo commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/DeploymentArtifactType.java:
```diff
@@ -0,0 +1,47 @@
+
+  @Override
+  public String toString() {
+    return "." + name().toLowerCase();
```
toLowerCase should have a Locale.US

elharo commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/DoManagedVmDeployment.java:
```diff
@@ -0,0 +1,159 @@
+  }
+
+  public void run() {
+    Preconditions.checkState(deploymentArtifactPath != null);
```
checkNotNull

elharo commented on google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/DoManagedVmDeployment.java:
```diff
@@ -0,0 +1,159 @@
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    try {
```
Looks like you coudl combine these two try blocks

etanshaul commented on google-cloud-tools-plugin/src/com/google/gct/idea/CloudToolsPluginInitializationComponent.java:
```diff
@@ -50,6 +52,9 @@ public void initComponent() {
               ConfigurationType.CONFIGURATION_TYPE_EP, new CloudDebugConfigType());
     }
+    if (pluginInfoService.shouldEnable(GctFeature.MANAGEDVM)) {
+      pluginConfigurationService.registerExtension(ServerType.EP_NAME, new ManagedVmCloudType());
```
In order to create a deployment configuration when trying to deploy to MVM I had to create a `DeployToServerConfigurationType` extension with name `CONFIGURATION_TYPE_EP` here in the initialization and pass the `ManangedVmCloudType` to it. 

IJ 15.0.3 now throws an exception if this isn't set up (instead of previously returning null): see `DeployToServerConfigurationRegsitrar.getDeployConfigurationType`.

patflynn replied:
Aha, all it required in the plugin.xml was the one server type registration, so when I tested and saw the MVM Cloud show up in the settings, I assumed I was good to go.

I'll check this out.

etanshaul commented:
lgtm (pending the comments). I confirmed the feature locking works - MVM deployment is hidden from the user as is. 

We can go ahead an merge into master and work from there. The other option is to keep the PR open and work from this branch. I'm in favor of merging so that we can split future MVM related work into smaller PRs.

patflynn commented:
@elharo re: "Can you file the additional work as independent feature requests?"
Yes! That was the plan, doing that today.

patflynn commented:
@elharo  All points addressed. 

The 'browse' button issue was a bad label vs bad text. It's actually the text shown at the top of the dialog window, so I've changed they keys accordingly.

Thanks for noticing the multiple try catches in the deployment. I had meant to factor out that code but forgot to. Now I have.

@etanshaul thanks for pointing out the issue with the flagging. That's now fixed.

patflynn commented:
Just waiting on LGTM from @elharo and I'll merge this and go to the beach.
